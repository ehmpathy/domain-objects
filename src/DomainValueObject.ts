import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Value Object is a type of Domain Object for which:
 * - is immutable (i.e, properties do not change over time)
 * - is defined by its properties (i.e., it is uniquely identified by the composition of all of its properties)
 *
 * For example,
 * - A `Address { street, city, state, country }` is a value object. Changing any of the properties, like `street`, produces a completely new address
 * - A `Name { first, last }` is a value object. Changing either property means you are dealing with a new name.
 */
export abstract class DomainValueObject<T> extends DomainObject<T> {}
