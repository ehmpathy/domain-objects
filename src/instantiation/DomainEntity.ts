import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, an Entity is a type of Domain Object for which:
 * - properties change over time (e.g., it has a life cycle)
 * - identity matters (e.g., two entities could have the same properties, differing only by id, and still be valid)
 *
 * For example,
 * - A `User { uuid, name, address, profilePicture }` is an entity. Properties like `name`, `address`, and `profilePicture` can change over time.
 * - A `Job { uuid, requestId, status }` is an entity. We expect `status` to change over time as the job goes from, for example, `QUEUED` to `FULFILLED`.
 */
export class DomainEntity<T> extends DomainObject<T> {
  /**
   * `DomainEntity.unique` defines all of the properties of the entity that the entity is naturally unique on
   *
   * For example,
   * - a `FileProcessingJob { uuid, filePath, status }` is likely going to be unique on the `filePath` if we only ever need to process the same file once.
   * - a `SupportTicket { uuid, userId, type, status }`, on the other hand, is likely only going to be unique on the `uuid`, because the same user can create the same support ticket over and over again
   */
  public static unique: readonly string[];

  /**
   * `DomainEntity.updatable` defines all of the properties of the entity that are updatable
   *
   * For example,
   * - a `User { uuid, name, address, picture }` would likely have `unique = ['name', 'address', 'picture']`, to enable the user to update their information as it changes over time
   * - an `Email { uuid, externalId, toAddress, fromAddress, message, status }` is likely only going to have `status` be updatable, as by changing the `toAddress` would, in this case, make us consider it to be a different email!
   */
  public static updatable: readonly string[];
}
