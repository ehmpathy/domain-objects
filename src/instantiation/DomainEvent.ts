import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, an Event is a type of Domain Object for which:
 * - properties are immutable (e.g., something happened)
 * - identity does matter
 *   - i.e., it is uniquely identified by what happened and to what
 *   - i.e., if you change the uniquely identifying properties (when or to what) of an event, it will be considered a new event
 *   - i.e., not all properties on the event uniquely identify the event
 *
 * The purpose of a Domain Event is to represent interesting things that happen in the domain.
 *
 * For example,
 * - An `ItemViewedEvent { itemUuid, customerUuid, occurredAt }` is an event. Changing when it `occurredAt` which `customerUuid` did it or which `itemUuid` it happened to would result in a new event.
 * - A `PowerSupplyMeasuredEvent { batteryUuid, current, voltage, temperature, occurredAt }` is an event. Changing when it `occurredAt` or which `batteryUuid` it occurred to would result in a new event. Note: changing the `current`, `voltage`, or `temperature` would _not_ result in a new event though, since only one measurement can be made at a time for a `batteryUuid`
 */
export abstract class DomainEvent<
  T extends Record<string, any>,
> extends DomainObject<T> {
  /**
   * `DomainEvent.unique` defines all of the properties of the event that the event is naturally unique on.
   *
   * It should always include a timestamp.
   *
   * For example,
   * - a `FileOpenedEvent { fileUuid, userUuid, occurredAt }` is likely going to be unique on the all of the keys: the combination of `['fileUuid', 'userUuid', 'occurredAt']` uniquely identifies the event.
   * - an `AirComfortMeasuredEvent { locationUuid, sensorUuid, temperature, humidity, occurredAt }`, on the other hand, is likely only going to be unique on `['locationUuid', 'sensorUuid', 'occurredAt']`, because logically we cant have a measurement from the same `sensor` + `location` at the same `time`
   */
  public static unique: readonly string[];
}
