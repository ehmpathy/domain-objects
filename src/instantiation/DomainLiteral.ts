import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Literal (a.k.a. Value Object), is a type of Domain Object for which:
 * - properties are immutable
 *   - i.e., it represents some literal value which happens to have a structured object shape
 *   - i.e., if you change the value of any of its properties, it is a different literal
 * - identity does not matter
 *   - i.e., it is uniquely identifiable by its non-metadata properties
 *
 * The purpose of a Domain Literal is to represent distinct, literal values which happen to have a structured object shape
 *
 * For example,
 * - A `Address { street, city, state, country }` is a literal. Changing any of the properties, like `street`, produces a completely new address
 * - A `Geocode { latitude, longitude }` is a literal. Changing either property means you are dealing with a new geocode
 */
export abstract class DomainLiteral<
  T extends Record<string, any>,
> extends DomainObject<T> {}
