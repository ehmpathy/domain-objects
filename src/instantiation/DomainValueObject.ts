import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Value Object is a type of Domain Object for which:
 * - properties are immutable (e.g., its a fact)
 * - identity does not matter
 *   - i.e., it is uniquely identifiable by its properties
 *   - i.e., if you change the value of any of its properties, it is now considered a new value-object
 *
 * For example,
 * - A `Address { street, city, state, country }` is a value object. Changing any of the properties, like `street`, produces a completely new address
 * - A `Name { first, last }` is a value object. Changing either property means you are dealing with a new name.
 */
export abstract class DomainValueObject<T> extends DomainObject<T> {}
