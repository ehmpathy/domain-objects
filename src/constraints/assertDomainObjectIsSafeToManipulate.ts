import { DomainObject } from '../instantiation/DomainObject';

export class DomainObjectNotSafeToManipulateError extends Error {
  constructor({ unsafeKeys, className }: { unsafeKeys: string[]; className: string }) {
    const message = `
DomainObject '${className}' is not safe to manipulate.

Specifically, the keys: ${JSON.stringify(unsafeKeys)}.

Please make sure all nested objects are DomainObjects and are explicitly defined on the class definition, using the 'nested' static property.

For example:
\`\`\`ts
  class ${className} ... {
    public static nested = { ${unsafeKeys[0]}: ... };
  }
\`\`\`
    `;
    super(message);
  }
}

/**
 * For internal use only. DomainObjects must have all nested domain objects defined in their `DomainObject.nested` definition. Otherwise, we could be dealing with a domain object that is not instantiated, and thus we will not appropriately serialize it.
 *
 * Therefore, if any DomainObject has nested objects that are not instantiated as DomainObjects, there is a risk that the nested object is really a nested, uninstantiated DomainObject - which would cause bugs if uncaught.
 */
export const assertDomainObjectIsSafeToManipulate = <T extends Record<string, any>>(obj: DomainObject<T> & Record<string, any>): void => {
  // grab all the keys that have objects defined for their values
  const nestedObjectKeys = Object.keys(obj).filter((key) => typeof obj[key] === 'object' && obj[key] !== null);

  // grab all the "nested object keys" who's values are defined as "should be domain object"
  const nestedDomainObjectKeysDefined = Object.keys((obj.constructor as typeof DomainObject).nested ?? {});
  const nestedNonDomainObjectObjectKeys = nestedObjectKeys.filter((key) => !nestedDomainObjectKeysDefined.includes(key)); // not explicitly defined as a nested domain object key

  // now apply some filters to those concerning keys
  let concerningKeys = nestedNonDomainObjectObjectKeys;

  // now filter out the keys which correspond to well known objects that we known are serializable and are not potentially DomainObjects (e.g., dates)
  concerningKeys = concerningKeys.filter((concerningKey) => {
    const value = obj[concerningKey];
    if (value instanceof Date) return false; // if its a date, not concerning
    if (value instanceof Buffer) return false; // if its a buffer, not concerning
    return true; // if it wasn't filtered out by now, its still concerning
  });

  // now filter out the keys which correspond to arrays of objects which are safe to manipulate; // TODO: consider how we can move this check to not be dependent on runtime contents of the domain object instance
  concerningKeys = concerningKeys.filter((concerningKey) => {
    const value = obj[concerningKey];
    if (!Array.isArray(value)) return true; // if its not an array, its still concerning
    try {
      assertDomainObjectIsSafeToManipulate(value);
      return false; // if we reached here, that means that the contents of the array _are_ safe, so not concerning
    } catch (error) {
      if (error instanceof DomainObjectNotSafeToManipulateError) return true; // if it was this error, that means that this key is definitely still concerning
      throw error; // otherwise its an error we couldn't handle, so we should pass it up
    }
  });

  // if we have any concerningKeys still, then its not safe!
  if (concerningKeys.length) throw new DomainObjectNotSafeToManipulateError({ unsafeKeys: concerningKeys, className: obj.constructor.name });
};
