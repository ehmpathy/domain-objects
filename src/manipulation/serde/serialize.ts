/* eslint-disable @typescript-eslint/no-use-before-define */

/* eslint-disable no-underscore-dangle */
import { assertDomainObjectIsSafeToManipulate } from '../../constraints/assertDomainObjectIsSafeToManipulate';
import { DomainEntity } from '../../instantiation/DomainEntity';
import { DomainLiteral } from '../../instantiation/DomainLiteral';
import { DomainObject } from '../../instantiation/DomainObject';
import { getUniqueIdentifier } from '../getUniqueIdentifier';

interface SerializeOptions {
  /**
   * set this to true if you need to deserialize the domain objects losslessly
   *
   * note
   * - this will make it so we do not call `getUniqueIdentifier` on the nested domain objects
   * - by default, this is set to false, so that this can be used for change detection
   */
  lossless?: boolean;

  /**
   * set this to true if the order of items in arrays does not matter in your usecase
   *
   * note
   * - this will make it so that the original order of items in arrays is not preserved
   * - by default, this is set to false, since order matters in arrays by common definition
   */
  orderless?: boolean;
}

/**
 * Returns a deterministic, human readable, and de-serializable string representation of a Domain Object.
 *
 * Deterministic:
 * - orders object keys
 * - sorts elements in arrays
 * - when lossless=false, replaces nested domain objects with their minimal and static representation (i.e., calls `getUniqueIdentifier` on any nested domain objects)
 *
 * Use Cases:
 * - change detection
 *   - e.g., determine whether there has been a change to an entity or not, by serializing the previous and current versions, and comparing equality
 *   - i.e., `const areEqual = serialize(omitMetadataValues(userBefore)) === serialize(omitMetadataValues(userNow));`
 *
 * - identity comparison
 *   - e.g., determine whether two domain objects are the same, by getting their unique identities, serializing each and comparing equality
 *   - i.e., `const areEqual = serialize(getUniqueIdentity(contactMethod1)) === serialize(getUniqueIdentity(contactMethod2))`
 *
 * - persistance
 *   - e.g., serialize domain objects into a persistant store in string format (to be later revived with the `deserialize` method)
 */
export const serialize = (
  value: any,
  options: SerializeOptions = { lossless: false, orderless: false },
): string => {
  // convert the value to a deterministically serializable object
  const serializableValue = toSerializable(value, options, true);

  // json stringify it last, so that we dont have a recursively growing set of nested quotes
  return JSON.stringify(serializableValue);
};

/**
 * helper method for serialize
 *
 * converts any value to a deterministically serializable representation (i.e., sorts arrays + object keys, casts .toString(), handles domain objects, etc)
 */
const toSerializable = (
  value: any,
  options: SerializeOptions,
  root?: boolean,
): any => {
  // if this value is not an array and is not an object, then it's a literal, so no more preparation required, return it itself
  if (!Array.isArray(value) && typeof value !== 'object') return value;

  // if its null, return it too (null is typeof 'object' in js :shrug:)
  if (value === null) return null;

  // if its an array, then `toSerializable` each element, and sort the result if needed
  if (Array.isArray(value))
    return value
      .map((el) => toSerializable(el, options))
      .sort((a, b) => {
        // if order was said to not matter, then sort in ascending value of their stringified form, to allow for deterministic comparisons
        if (options.orderless)
          return JSON.stringify(a) < JSON.stringify(b) ? -1 : 1;

        // otherwise, retain original order
        return 0;
      });

  // if it has a `.toString()` method, use the `.toString()` method; (e.g., new Date().toString())
  if (value.toString()) {
    const toStringResult = value.toString();
    if (toStringResult !== '[object Object]') return toStringResult; // use the result, unless its the "generic object -> ['object Object']" result, which isn't helpful
  }

  // otherwise, prepare to make the object stringifiable
  return toSerializableObject(value, options, root);
};

/**
 * helper method for `toSerializable`
 *
 * converts all objects to deterministically serializable objects (e.g., sort keys, handle domain objects, etc)
 */
const toSerializableObject = (
  obj: Record<string, any>,
  options: SerializeOptions,
  root?: boolean,
) => {
  const stringifiableObj: Record<string, any> = {};

  // if this object is a DomainObject, add its name to the serial parts as metadata (i.e., underscore suffix)
  if (obj instanceof DomainObject)
    stringifiableObj._dobj = obj.constructor.name;

  // if this object is a domain object, make sure that it is safe to manipulate
  if (obj instanceof DomainObject) assertDomainObjectIsSafeToManipulate(obj);

  // if this object is a persistable DomainObject AND its not the root object we're serializing && lossless is off, then only consider its unique properties for serialization; (i.e., only consider which domain objects the root object references, not the current state of the domain objects the root object references)
  let objToMakeSerializable = obj;
  if (
    !root &&
    !options.lossless &&
    (obj instanceof DomainEntity || obj instanceof DomainLiteral)
  ) {
    objToMakeSerializable = getUniqueIdentifier(obj);
  }

  // for each key, in sorted order, get the serial representation
  Object.keys(objToMakeSerializable)
    .sort()
    .forEach((key) => {
      stringifiableObj[key] = toSerializable(obj[key], options);
    });

  // return it
  return stringifiableObj;
};
