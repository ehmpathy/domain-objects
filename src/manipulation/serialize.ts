/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable no-underscore-dangle */
import { assertDomainObjectIsSafeToManipulate } from '../constraints/assertDomainObjectIsSafeToManipulate';
import { DomainEntity } from '../instantiation/DomainEntity';
import { DomainObject } from '../instantiation/DomainObject';
import { DomainValueObject } from '../instantiation/DomainValueObject';
import { getUniqueIdentifier } from './getUniqueIdentifier';

/**
 * Returns a deterministic, human readable, and de-serializable string representation of a Domain Object.
 *
 * Deterministic:
 * - orders object keys
 * - sorts elements in arrays
 * - replaces nested domain objects with their minimal and static representation (i.e., calls `getUniqueIdentifier` on any key's value w/ instance of DomainEntity | DomainValueObject)
 *
 * Use Cases:
 * - change detection
 *   - e.g., determine whether there has been a change to an entity or not, by serializing the previous and current versions, and comparing equality
 *   - i.e., `const areEqual = serialize(userBefore) === serialize(userNow);`
 *
 * - identity comparison
 *   - e.g., determine whether two domain objects are the same, by getting their unique identities, serializing each and comparing equality
 *   - i.e., `const areEqual = serialize(getUniqueIdentity(contactMethod1)) === serialize(getUniqueIdentity(contactMethod2))`
 */
export const serialize = (value: any): string => {
  // 1. convert the value to a deterministically serializable object
  const serializableValue = toSerializable(value, true);

  // 2. json stringify it last, so that we dont have a recursively growing set of nested quotes
  return JSON.stringify(serializableValue);
};

/**
 * helper method for serialize
 *
 * converts any value to a deterministically serializable representation (i.e., sorts arrays + object keys, casts .toString(), handles domain objects, etc)
 */
export const toSerializable = (value: any, root?: boolean): any => {
  // if this value is not an array and has a `.toString` definition, return it itself
  if (!Array.isArray(value) && typeof value !== 'object') return value;

  // if its null, return it too (null is typeof 'object' in js :shrug:)
  if (value === null) return null;

  // if its an array, then `toSerializable` each element, and sort the result
  if (Array.isArray(value)) return value.map((el) => toSerializable(el)).sort((a, b) => (JSON.stringify(a) < JSON.stringify(b) ? -1 : 1)); // stringify to sort, to sort objects correctly

  // if it has a `.toString()` method, use the `.toString()` method; (e.g., new Date().toString())
  if (value.toString()) {
    const toStringResult = value.toString();
    if (toStringResult !== '[object Object]') return toStringResult; // use the result, unless its the "generic object -> ['object Object']" result, which isn't helpful
  }

  // otherwise, prepare to make the object stringifiable
  return toSerializableObject(value, root);
};

/**
 * helper method for `toSerializable`
 *
 * converts all objects to deterministically serializable objects (e.g., sort keys, handle domain objects, etc)
 */
const toSerializableObject = (obj: Record<string, any>, root?: boolean) => {
  const stringifiableObj: Record<string, any> = {};

  // if this object is a DomainObject, add its name to the serial parts as metadata (i.e., underscore suffix)
  if (obj instanceof DomainObject) stringifiableObj._dobj = obj.constructor.name;

  // if this object is a domain object, make sure that it is safe to manipulate
  if (obj instanceof DomainObject) assertDomainObjectIsSafeToManipulate(obj);

  // if this object is a persistable DomainObject AND its not the root object we're serializing, then only consider its unique properties for serialization; (i.e., only consider which domain objects the root object references, not the current state of the domain objects the root object references)
  let objToMakeSerializable = obj;
  if (!root && (obj instanceof DomainEntity || obj instanceof DomainValueObject)) {
    objToMakeSerializable = getUniqueIdentifier(obj);
  }

  // for each key, in sorted order, get the serial representation
  Object.keys(objToMakeSerializable)
    .sort()
    .forEach((key) => {
      stringifiableObj[key] = toSerializable(obj[key]);
    });

  // return it
  return stringifiableObj;
};
