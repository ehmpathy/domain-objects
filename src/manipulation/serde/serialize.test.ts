import uuid from 'uuid';
import { DomainObject } from '../..';

import { DomainObjectNotSafeToManipulateError } from '../../constraints/assertDomainObjectIsSafeToManipulate';
import { DomainEntity } from '../../instantiation/DomainEntity';
import { DomainValueObject } from '../../instantiation/DomainValueObject';
/* eslint-disable no-useless-escape */
import { serialize } from './serialize';

describe('serialize', () => {
  describe('basic types', () => {
    it('should serialize strings', () => {
      const serial = serialize('hello!');
      expect(serial).toEqual('"hello!"');
    });
    it('should serialize numbers', () => {
      const serial = serialize(821);
      expect(serial).toEqual('821');
    });
    it('should serialize dates', () => {
      const serial = serialize(new Date('2020-08-21 00:00:00'));
      expect(serial).toContain('Fri Aug 21 2020 00:00:00');
    });
    it('should serialize undefined', () => {
      const serial = serialize(undefined);
      expect(serial).toEqual(undefined);
    });
    it('should serialize nulls', () => {
      const serial = serialize(null);
      expect(serial).toEqual('null');
    });
    it('should serialize buffers', () => {
      const serial = serialize(Buffer.from('821'));
      expect(serial).toEqual('"821"');
    });
  });
  describe('arrays', () => {
    it('should be able to serialize arrays', () => {
      const serial = serialize(['821', 721, 'leopard', 7, 'apple', 3]);
      expect(serial).toEqual(`[\"821"\,\"apple\",\"leopard\",3,7,721]`); // note that it sorts them
    });
    it('should serialize them deterministically - independent of order of values', () => {
      const serialA = serialize([3, '821', 721, 'leopard', 'apple']);
      const serialB = serialize([721, 'leopard', 3, 'apple', '821']);
      expect(serialB).toEqual(serialA); // should be equivalent, since it serializes deterministically by value (due to sorting)
    });
    it('should serialize arrays even if they have objects - still deterministically', () => {
      const serialA = serialize([
        'banana',
        { id: 1, value: 821, meaning: 42 },
        821,
        { id: 0, value: undefined, meaning: null }, // should go first, because id is 0
      ]);
      const serialB = serialize([
        { id: 0, value: undefined, meaning: null }, // should go first, because id is 0
        { id: 1, value: 821, meaning: 42 },
        821,
        'banana',
      ]);
      expect(serialB).toEqual(serialA); // should be equivalent
    });
  });
  describe('objects', () => {
    it('should be able to serialize an object with all sorts of types', () => {
      const serial = serialize({
        color: 'blue',
        cost: 821,
        orders: [
          { id: 1, value: 821, meaning: 42 },
          { id: 0, value: undefined, meaning: null },
        ],
        application: {
          type: 'PAINTING',
        },
      });
      expect(serial).toEqual(
        // note that it sorts them
        `{"application":{"type":"PAINTING"},"color":"blue","cost":821,"orders":[{"id":0,"meaning":null},{"id":1,"meaning":42,"value":821}]}`,
      );
    });
    it('should sort object keys, for determinism', () => {
      const serialA = serialize({
        color: 'blue',
        application: 821,
      });
      const serialB = serialize({
        application: 821,
        color: 'blue',
      });
      expect(serialA).toEqual(serialB); // should show that they're equivalent, due to sorting
    });
  });
  describe('domain objects', () => {
    // define some domain objects to test with
    interface Spaceship {
      serialNumber: string;
      fuelQuantity: number;
      passengers: number;
    }
    class Spaceship extends DomainEntity<Spaceship> implements Spaceship {
      public static unique = ['serialNumber'];
      public static updatable = ['serialNumber'];
    }
    interface Address {
      id?: number;
      galaxy: string;
      solarSystem: string;
      planet: string;
      continent: string;
    }
    class Address extends DomainValueObject<Address> implements Address {}
    interface Spaceport {
      uuid: string;
      address: Address;
      spaceships: Spaceship[];
    }
    class Spaceport extends DomainEntity<Spaceport> implements Spaceport {
      public static unique = ['uuid'];
      public static updatable = ['spaceships'];
      public static nested = { address: Address, spaceships: Spaceship };
    }

    // run the tests
    it('should add the domain object name as metadata', () => {
      const ship = new Spaceship({
        serialNumber: '__UUID__',
        fuelQuantity: 9001,
        passengers: 21,
      });
      const serial = serialize(ship);
      expect(serial).toContain('Spaceship');
      expect(serial).toEqual(`{\"_dobj\":\"Spaceship\",\"fuelQuantity\":9001,\"passengers\":21,\"serialNumber\":\"__UUID__\"}`);
    });
    it('should add the domain object name as metadata recursively, for all nested domain objects too', () => {
      const shipA = new Spaceship({
        serialNumber: '__SHIP_A__',
        fuelQuantity: 9001,
        passengers: 21,
      });
      const shipB = new Spaceship({
        serialNumber: '__SHIP_B__',
        fuelQuantity: 7000,
        passengers: 42,
      });
      const spaceport = new Spaceport({
        uuid: '__SPACEPORT_UUID__',
        address: new Address({ galaxy: 'Milky Way', solarSystem: 'Sun', planet: 'Earth', continent: 'North America' }),
        spaceships: [shipA, shipB],
      });
      const serial = serialize(spaceport);

      // check that it has the metadata on there, specifying which domain object each object was
      expect(serial).toContain('Spaceship');
      expect(serial).toContain('Address');
      expect(serial).toContain('Spaceport');

      // and then record a snapshot, just so we can see how it looks as a visual example
      expect(JSON.parse(serial)).toMatchSnapshot();
    });
    it('should replace each nested domain object with its unique representation, so that updates to the non-unique properties of referenced entities dont affect comparisons', () => {
      const shipA = new Spaceship({
        serialNumber: '__SHIP_A__',
        fuelQuantity: 9001,
        passengers: 21,
      });
      const shipB = new Spaceship({
        serialNumber: '__SHIP_B__',
        fuelQuantity: 7000,
        passengers: 42,
      });
      const spaceport = new Spaceport({
        uuid: '__SPACEPORT_UUID__',
        address: new Address({ id: 821, galaxy: 'Milky Way', solarSystem: 'Sun', planet: 'Earth', continent: 'North America' }),
        spaceships: [shipA, shipB],
      });
      const serialA = serialize(spaceport);

      // now define spaceport again, but this time the fuel quantity of shipB has decreased (since it was flying)
      spaceport.spaceships[1].fuelQuantity = 4200;
      const serialB = serialize(spaceport);

      // we should not be tracking "fuelQuantity" or "passengers", because those properties are updatable and in domain modeling, we only care about which domain objects are being referenced (i.e., bounded scope of consideration) - not what state referenced object's updatable properties are in; that's why we only care about spaceship.serialNumber here, since that uniquely identifies the spaceship being referenced
      expect(serialA).not.toContain('fuelQuantity');
      expect(serialA).not.toContain('passengers');

      // note, we should still be tracking all parts of the address (except id) - since those are required to uniquely identify the address
      expect(serialA).toContain('galaxy');
      expect(serialA).toContain('solarSystem');
      expect(serialA).toContain('planet');
      expect(serialA).toContain('continent');

      // their serials should be equivalent, because, due to domain modeling, we know that those two space ports are still the same (i.e., they reference the same ships, even though the ship's updatable properties have changed!)
      expect(serialB).toEqual(serialA);
    });

    interface GlowWorm {
      glowing: boolean;
      color: string;
    }
    class GlowWorm extends DomainObject<GlowWorm> implements GlowWorm {}
    it('should not throw an error trying to serialize a safe generic domain object', () => {
      const worm = new GlowWorm({
        glowing: true,
        color: 'purple',
      });
      const serial = serialize(worm);
      expect(serial).toContain('GlowWorm');
      expect(serial).toEqual(`{\"_dobj\":\"GlowWorm\",\"color\":\"purple\",\"glowing\":true}`);
    });

    describe('safety', () => {
      it('should throw an error if domain object is not safe to manipulate', () => {
        interface PlaneManufacturer {
          name: string;
        }
        interface Plane {
          uuid: string;
          manufacturer: PlaneManufacturer;
          passengerLimit: number;
        }
        class Plane extends DomainEntity<Plane> implements Plane {
          public static unique = ['uuid'];
        }
        const phone = new Plane({
          uuid: uuid(),
          manufacturer: { name: 'boeing' },
          passengerLimit: 821,
        });
        try {
          serialize(phone);
        } catch (error) {
          expect(error).toBeInstanceOf(DomainObjectNotSafeToManipulateError);
        }
      });
    });
  });
});
