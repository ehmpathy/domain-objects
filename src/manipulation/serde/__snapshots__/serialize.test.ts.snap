// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialize domain objects should add the domain object name as metadata recursively, for all nested domain objects too 1`] = `
{
  "_dobj": "Spaceport",
  "address": {
    "_dobj": "Address",
    "continent": "North America",
    "galaxy": "Milky Way",
    "planet": "Earth",
    "solarSystem": "Sun",
  },
  "spaceships": [
    {
      "_dobj": "Spaceship",
      "serialNumber": "__SHIP_A__",
    },
    {
      "_dobj": "Spaceship",
      "serialNumber": "__SHIP_B__",
    },
  ],
  "uuid": "__SPACEPORT_UUID__",
}
`;
